### PR TITLE
Refactored `precommit.py` into tox

### PR DESCRIPTION
**Pull request by altendky**
Draft for:
- [x] Self review
- [x] ~~Maybe make `tox -e test` work (that is, without `-py39` etc)~~
- [x] Update documentation

This patch refactors the logic from `precommit.py` into `tox.ini` so
that the steps can be run in isolation in the continuous integration.

### CODE REVIEW

**Comment by altendky** [comment]
Welp, there's a green check...  I'll look over it tomorrow and see what silliness is floating around here.

**Comment by altendky** [comment]
Alrighty.  At present, the equivalent of `./precommit.py` would be `tox -e check,test-py39` and for `./precommit.py --overwrite` it might be `tox -e format,check,test-py39`.  They are not exactly the same nor is this some final 'this is proper' claim.  Just where it is at now.  I dislike having to specify the Python version for test vs. it just defaulting to something.  `precommit.py`, or another file, could be written to wrap up these calls if you want.  Or let developers just deal with a little change.

The separation between the checks and the tests is because checks are generally run against the source and tests are nice to run against the installed code, in CI anyways.  If you want to have less focus on that separation for local use, we can look into good ways to mash them together for developers.

I took the liberty of expanding some check coverage since it required only minimal changes.  That can certainly be rolled back if preferred.  Or, if wanted, there may be some more expansion that could be done.

Anyways, maybe take a quick pass and let me know how you feel about the big picture.  I can try to respond to that before either of us get our heads down into too many details.  Or just go straight to the details, whatever.  `:]`

**Comment by altendky** [comment]
I just adjusted so that running just `tox` will execute the `check` and `test-py39` environments.  That default could be easily changed according to your preference.

**Comment by altendky** [comment]
@mristin, I don't want to nag but figured I'd see if you wanted to weigh in on the direction this is going or if I should plow ahead into the docs.  Would you want to keep `precommit.py` around or remove it in favor of having just one recommended way?

File: tox.ini [hunk 1]
```diff
@@ -1,17 +1,43 @@
 [tox]
-envlist = py35,py36
+envlist = check,test-py39
```
Comment (jugmac00) [inline tox.ini:2]:
```suggestion
envlist = check,py39,py38,py37,py36...
```

File: tox.ini [hunk 2]
```diff
@@ -1,17 +1,43 @@
...
+    dev
 setenv =
+    _PYTHON_SOURCE_PATHS = {toxinidir}/bin/pylddwrap {toxinidir}/lddwrap/ {toxinidir}/precommit.py {toxinidir}/setup.py
+    _PYTHON_TEST_PATHS = {toxinidir}/tests/
+    _PYTHON_PATHS = {env:_PYTHON_SOURCE_PATHS} {env:_PYTHON_TEST_PATHS}
+    _YAPF_OPTIONS = --style={toxinidir}/style.yapf --recursive {env:_PYTHON_PATHS}
+    _ISORT_OPTIONS = --settings-path={toxinidir}/.isort.cfg --recursive {env:_PYTHON_PATHS}
     COVERAGE_FILE={envbindir}/.coverage
 
+[testenv:test-py{35,36,37,38,39,py36,py37}]
```
Comment (jugmac00) [inline tox.ini:16]:
```suggestion
```

**Comment by mristin** [comment]
Hi @altendky ,

> @mristin, I don't want to nag but figured I'd see if you wanted to weigh in on the direction this is going or if I should plow ahead into the docs. Would you want to keep `precommit.py` around or remove it in favor of having just one recommended way?

Please apologize, I missed this message in my mailbox since it was surrounded by too many commit messages. 

I'd prefer to keep `precommit.py` and `precommit.py --overwrite` which just calls the two commands you mentioned. That way it's very easy for the developers to get started (and also to prevent errors, *.e.g.*, a missed check locally due to a minor typo).

I reviewed the changes so far and all looks good to me.

**Comment by mristin** [comment]
@altendky please also have a look at the change of the title, and feel free to change it to a more approapriate one.

**Comment by altendky** [comment]
Title is fine by me, and certainly more descriptive.  `:]`  I'll update `precommit.py` accordingly.  And yeah...  easy to miss stuff.  I personally gave up on the emails a couple months back and just use https://github.com/notifications these days.  But that only works well if you are busy on GitHub anyways.

**Comment by altendky** [comment]
Got that bit I think.  Anything else to improve?

**Comment by mristin** [comment]
@altendky Looks good to me! Should I merge?

**Comment by altendky** [comment]
Go for it, thanks.  I'll catch up the others after.

**Comment by mristin** [comment]
This patch refactors the logic from precommit.py into tox.ini so
that the steps can be run in isolation in the continuous integration.

**Comment by mristin** [comment]
Thanks again, @altendky!

**Pull request merged**